packages/ens-utils/src/ethregistrar.ts

```diff
@@ -155,3 +179,284 @@ export function buildNFTRefFromENSName(
+  atTimestamp: Timestamp,
+  expirationTimestamp: Timestamp,
+): Price {
+  const releasedTimestamp = addSeconds(expirationTimestamp, GRACE_PERIOD);
+  const secondsSinceRelease = atTimestamp.time - releasedTimestamp.time;
+  if (secondsSinceRelease < 0) {
+    /* if as of the moment of `atTimestamp` a name hasn't expired yet then there is no temporaryPremium */
+    return {
+      value: 0n,
+      currency: Currency.Usd,
+    };
+  }
+
+  const fractionalDaysSinceRelease =
+    Number(secondsSinceRelease) / Number(SECONDS_PER_DAY.seconds);
+
+  const decayFactor = PREMIUM_DECAY ** fractionalDaysSinceRelease;
+
+  const decayedPrice = approxScalePrice(PREMIUM_START_PRICE, decayFactor);
+  const offsetDecayedPrice = subtractPrices(decayedPrice, PREMIUM_OFFSET);
+
+  /* the temporary premium can never be less than $0.00 */
+  if (offsetDecayedPrice.value < 0n) {
+    return {
+      value: 0n,
+      currency: Currency.Usd,
+    };
+  }
+
+  return offsetDecayedPrice;
+}
+
+/*
+  This is an "internal" helper function only. It can't be directly used anywhere else because
+  it is too easy to accidently not include the registration object when it should be passed.
+  Three different functions are created right below this one, which are the ones that are
+  safe to be used across the platform, and are then, the ones being exported.
+*/
+const AvailableNamePriceUSD = (
+  parsedName: DomainName,
+  registerForYears = DEFAULT_REGISTRATION_YEARS,
+  registration: Registration | null = null,
+  additionalFee: Price | null = null,
+): Price | null => {
+  if (!parsedName.normalizedName) return null;
+
+  const defaultPrice: Readonly<Price> = {
+    value: 500n,
+    currency: Currency.Usd,
+  };
+  const shortNamePremium: Record<number, Readonly<Price>> = {
+    [MIN_ETH_REGISTRABLE_LABEL_LENGTH]: {
+      value: 64000n,
+      currency: Currency.Usd,
+    },
+    4: {
+      value: 16000n,
+      currency: Currency.Usd,
+    },
+  };
+  const basePrice = shortNamePremium[parsedName.labelName.length]
+    ? shortNamePremium[parsedName.labelName.length]
+    : defaultPrice;
+
+  const namePriceForYears = multiplyPriceByNumber(
+    basePrice,
+    Number(registerForYears),
+  );
+
+  const namehashPrice = additionalFee
+    ? addPrices([additionalFee, namePriceForYears])
+    : namePriceForYears;
+
+  if (registration) {
+    const premiumPrice = nameCurrentTemporaryPremium(registration);
+
+    return premiumPrice
+      ? addPrices([premiumPrice, namehashPrice])
+      : namehashPrice;
+  }
+
+  return namehashPrice;
+};
+
+const DEFAULT_REGISTRATION_YEARS = 1;
+
+/*
+  Below function returns the "timeless" price for a name, that takes no consideration
+  of the current status of the name. This is useful for various cases, including in
+  generating messages that communicate how much a name costs to renew, how much
+  a name will cost at the end of a premium period, etc..
+*/
+export const AvailableNameTimelessPriceUSD = (
```

Comment:
It will be awesome if you can create unit tests for this :+1:

Reply:
I have already created the description of the essential test cases,

OK if I add its logics once `mock-utils` with `getENSNameForVariant` is available?